trace-utils/src/msgpack_decoder/decode/span_event.rs

```diff
@@ -0,0 +1,411 @@
+        }
+        AttributeArrayKey::BoolValue => {
+            let boolean = read_boolean_bytes(buf);
+            if let Ok(value) = boolean {
+                match value {
+                    true => Ok(AttributeArrayValueBytes::Boolean(true)),
+                    false => Ok(AttributeArrayValueBytes::Boolean(false)),
+                }
+            } else {
+                Err(DecodeError::InvalidType("Invalid boolean field".to_owned()))
+            }
+        }
+        AttributeArrayKey::IntValue => {
+            Ok(AttributeArrayValueBytes::Integer(read_number_bytes(buf)?))
+        }
+        AttributeArrayKey::DoubleValue => {
+            Ok(AttributeArrayValueBytes::Double(read_number_bytes(buf)?))
+        }
+        _ => Err(DecodeError::InvalidFormat("Invalid attribute".to_owned())),
+    }
+}
+
+fn decode_attribute_array(
+    buf: &mut Bytes,
+    array_type: Option<u8>,
+) -> Result<AttributeArrayValueBytes, DecodeError> {
+    let mut attribute: Option<AttributeArrayValueBytes> = None;
+    let attribute_size =
+        rmp::decode::read_map_len(unsafe { buf.as_mut_slice() }).map_err(|_| {
+            DecodeError::InvalidType("Unable to get map len for attribute size".to_owned())
+        })?;
+
+    if attribute_size != 2 {
+        return Err(DecodeError::InvalidFormat(
+            "Invalid number of field for an attribute".to_owned(),
+        ));
+    }
+    let mut attribute_type: Option<u8> = None;
+
+    for _ in 0..attribute_size {
+        match read_string_ref(unsafe { buf.as_mut_slice() })?.parse::<AttributeArrayKey>()? {
+            AttributeArrayKey::Type => attribute_type = Some(read_number_bytes(buf)?),
+            key => attribute = Some(get_attribute_from_key(buf, key)?),
+        }
+    }
+
+    if let Some(value) = attribute {
+        if let Some(attribute_type) = attribute_type {
+            if type_from_attribute_array(&value) == attribute_type {
+                if let Some(array_type) = array_type {
+                    if array_type != attribute_type {
+                        return Err(DecodeError::InvalidType(
+                            "Array must have same type element".to_owned(),
+                        ));
+                    }
+                }
+                Ok(value)
+            } else {
+                Err(DecodeError::InvalidFormat(
+                    "No type for attribute".to_owned(),
+                ))
+            }
+        } else {
+            Err(DecodeError::InvalidType(
+                "Type mismatch for attribute".to_owned(),
+            ))
+        }
+    } else {
+        Err(DecodeError::InvalidFormat("Invalid attribute".to_owned()))
+    }
+}
+
+fn type_from_attribute_array(attribute: &AttributeArrayValueBytes) -> u8 {
+    match attribute {
+        AttributeArrayValueBytes::String(_) => 0,
+        AttributeArrayValueBytes::Boolean(_) => 1,
+        AttributeArrayValueBytes::Integer(_) => 2,
+        AttributeArrayValueBytes::Double(_) => 3,
```

Comment:
These values should be named constants. It complies to the same thing, but is a bit better for clarity and maintainability.

Reply:
Since the implementation lives in AttributeArrayValueBytes and the numbers don't have to be directly referenced elsewhere I don't think it's necessary.